core/trino-main/src/main/java/io/trino/server/security/oauth2/OAuth2Service.java

```diff
@@ -67,31 +83,59 @@
     private final TemporalAmount challengeTimeout;
     private final byte[] stateHmac;
 
+    private final String issuer;
+    private final List<String> allowedAudiences;
+
+    private final OAuth2TokenHandler tokenHandler;
+
+    private final boolean webUiOAuthEnabled;
+
     @Inject
-    public OAuth2Service(OAuth2Client client, @ForOAuth2 SigningKeyResolver signingKeyResolver, OAuth2Config oauth2Config)
+    public OAuth2Service(OAuth2Client client, @ForOAuth2 SigningKeyResolver signingKeyResolver, OAuth2Config oauth2Config, OAuth2TokenHandler tokenHandler, Optional<OAuth2WebUiInstalled> webUiOAuthEnabled)
             throws IOException
     {
         this.client = requireNonNull(client, "client is null");
         this.signingKeyResolver = requireNonNull(signingKeyResolver, "signingKeyResolver is null");
+        requireNonNull(oauth2Config, "oauth2Config is null");
 
         this.successHtml = Resources.toString(Resources.getResource(getClass(), "/oauth2/success.html"), UTF_8);
         this.failureHtml = Resources.toString(Resources.getResource(getClass(), "/oauth2/failure.html"), UTF_8);
         verify(failureHtml.contains(FAILURE_REPLACEMENT_TEXT), "login.html does not contain the replacement text");
 
-        requireNonNull(oauth2Config, "oauth2Config is null");
         this.scopes = oauth2Config.getScopes();
         this.challengeTimeout = Duration.ofMillis(oauth2Config.getChallengeTimeout().toMillis());
         this.stateHmac = oauth2Config.getStateKey()
                 .map(key -> sha256().hashString(key, UTF_8).asBytes())
                 .orElseGet(() -> secureRandomBytes(32));
+
+        this.issuer = oauth2Config.getIssuer();
+        this.allowedAudiences = ImmutableList.<String>builder()
+                .addAll(oauth2Config.getAdditionalAudiences())
+                .add(oauth2Config.getClientId())
+                .build();
+
+        this.tokenHandler = requireNonNull(tokenHandler, "tokenHandler is null");
+
+        this.webUiOAuthEnabled = requireNonNull(webUiOAuthEnabled, "webUiOAuthEnabled is null").isPresent();
+    }
+
+    public Response startOAuth2Challenge(UriInfo uriInfo)
+    {
+        return startOAuth2Challenge(uriInfo, Optional.empty());
     }
 
-    public OAuthChallenge startWebUiChallenge(URI callbackUri)
+    public Response startOAuth2Challenge(UriInfo uriInfo, String handlerState)
     {
-        Instant challengeExpiration = Instant.now().plus(challengeTimeout);
+        return startOAuth2Challenge(uriInfo, Optional.ofNullable(handlerState));
```

Comment:
`handlerState` should not be null, so use `Optional.of` here

Reply:
👍